### ISSUE TITLE
Revamp treatment of conflicted records

### ISSUE THREAD

**Issue by teixeirak** [forc-db/ForC#76]
I have just pushed changes with some new fields in MEASUREMENTS intended to deal with conflicted records. We need to fill these fields and figure out whether they work well. Here are the steps:
1. - [ ] @ValentineHerr, please generate script to fill these fields as described in the Metadata. Check with me if you have questions or if this new system is not working nicely. Please generate a list of records that need my attention.
2. - [ ] @teixeirak will then need to check records that need expert attention.
3. - [ ] Once this is done, we should be able to delete dup.type and dup.num.

**Comment by ValentineHerr** [comment]
@teixeirak,

In the made up case below, How would we fill in the conflict.type ?


ID | variable.names | notes | method.ID | citation.ID | conflicts | R.group | S.group | D.group | D.precedence | conflict.type
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
1 | A_C | W | 1 | A | D, R | 1 |   | 1 |   |  
2 | A_C | W | 1 | B | D |   |   | 1 |   | 
3 | A_C | W | 1 | A | D, R | 1 |   | 1 |   |  
4 | A_C | X | 1 | B | D |   |   | 1 |   |  
5 | A_C | Y | 2 | A | D |   |   | 1 |   |  
6 | A_C | Z | 2 | B | D |   |   | 1 |   |  
7 | A_OM | W | 1 | A | D, R | 2 |   | 1 |   |  
8 | A_OM | W | 1 | B | D |   |   | 1 |   | 
9 | A_OM | W | 1 | A | D, R | 2 |   | 1 |   |  
10 | A_OM | X | 1 | B | D |   |   | 1 |   |  
11 | A_OM | Y | 2 | A | D |   |   | 1 |   |  
12 | A_OM | Z | 2 | B | D |   |   | 1 |   |

**Comment by teixeirak** [comment]
I think they should all get "M,C". Does that make sense to you?

**Comment by ValentineHerr** [comment]
ok

**Comment by teixeirak** [comment]
Treat it as conflicting with all other measurements of that variable in that plot (S code).

**Comment by teixeirak** [comment]
Please treat these as duplicates. Note that your code is not listing instances where the date falls outside of the start.date-end.date range (17245 & 17269). However, that instance should be treated as a duplicate.

**Comment by ValentineHerr** [comment]
If the date is totally separate from the range of date it should be considered as duplicate ?

You are talking about S.group when you say duplicates in these cases right ?

**Comment by teixeirak** [comment]
If the date is totally separate from the range of date it should be considered as duplicate ? 
**In general, no. But in this particular case, it should be treated as a duplicate.**

You are talking about S.group when you say duplicates in these cases, right ?

**Actually, no, these cases are all duplicates because there's a 1-to-1 conflict. If there was a 1-to-many conflict, that would be S. Does that make sense/ seem consistent to you?**

**Comment by teixeirak** [comment]
Does that mean that I should not be looking at stand age at all ? 
**I'm not sure what you mean by this. Does it still apply given the answers above?**

**Comment by ValentineHerr** [comment]
Why are 17245 & 17269 a special case ?

I was thinking that the fact that 2 records would have different stand.age would be a reason to keep them as independent. I can understand that when there is a range of date that overlaps another record, we don't care about stand.age. But, for example, in a set of records with no dates at all but a stand.age different for each record. Should I still consider the set as "s" or should I look at stand.age to make them all independent?

**Comment by teixeirak** [comment]
Yes, please do consider stand.age.

In this case, the records come from two different studies that must apply different methods, as the difference in biomass is much larger than I'd expect over a small difference in time (2 years, if I remember correctly). Thus, the difference in methods outweighs the difference in years and makes them effective duplicates. I wouldn't want that to get counted as two comparable records and used to calculate change in biomass, for example. That said, technically, they should not be treated as duplicates, so maybe for consistency we should call them independent. Future code will need to ensure consistency when estimating changes in biomass (for example).

**Comment by ValentineHerr** [comment]
For those 1-to-1 conflict duplicates, should I give a "T" code ?

**Comment by teixeirak** [comment]
I think we've resolved this, correct? 1-to-1.date is D, 1-to-many.dates is S, and 1-to-NA is S (where NA gets capital S).

**Comment by ValentineHerr** [comment]
yes

**Comment by ValentineHerr** [comment]
For the example, If I follow the decision we made yesterday they would be independent because the start.date are different. Is that what we want ?

ID | measurement.ID | start.date | end.date
-- | -- | -- | --
1|3469|2002|2005
2|3493|2004|2006

**Comment by teixeirak** [comment]
If there is any overlap in the measurement period (as in the above example), it must be counted as a duplicate (with T as a conflict type).

**Comment by teixeirak** [comment]
I think that the only (or best) way to deal with this will be to allow multiple duplicate groups. 

In a case like this: ...
A  | 2001 - 2002
B  | 2001.5-2002.5
C | 2002-2003
...A and B are duplicates (D.group 1) and B and C are duplicates (D.group 2). B gets D.group=1,2.

~~
To clarify the questions above, by "date record", I assume you mean a date in the *date* field? And by "range records" you mean dates in *date.start* and *date.end* fields? If this is correct, the first thing to know is that for any flux, a single year in *date* should be interpreted as Jan 1-Dec 31 of that year. Then, the answers are as follows:

1.  a. Assuming we are not looking at an "S" relationship, treat each pair separately, as in the example above. An example might be:

ID | date | start.date | end.date | conflicts | D.group  
-- | -- | -- | -- | -- | -- 
A |NA|2001|2003|D|1,2
B  | 2001|NA | NA |D |1
C | NA|2002 | 2005 | D|2

or, adding complexity by including an "S" relationship: 

ID | date | start.date | end.date | conflicts|S.group | D.group  
-- | -- | -- | -- | -- | -- | --
A |NA|2001|2003|S,D|1|1
B  | 2001|NA | NA |s |1|NA
C  | 2002|NA | NA |s,D |1|2
D | NA|2002 | 2005 | D|NA|1,2

1b. This would be a D conflict.
2. Correct.

**Comment by ValentineHerr** [comment]
Ok,
I'll try to code that.

**Comment by ValentineHerr** [comment]
shouldn't A, C and D be in the same D.group (in the second example above) ?

**Comment by teixeirak** [comment]
Yes, corrected.

**Comment by ValentineHerr** [comment]
Ok I think I've got a running script now.
I need to double check a couple things on Monday but it shouldn't take too long.
I am closing this issue.

**Comment by ValentineHerr** [comment]
Hi Krista,
My code is pretty much good to go now but before I push the changes in Measurements, I wanted to double check a couple things with you:

1. The code does not handle the particular cases for BCI/conflict type P (but it does not over-right them). I don't know how to handle that (and I am not sure I really get the problem). Do you know any other site that would have that problem ? The code is currently grouping by site, plot and variable... maybe we can create a new column with the name of the plot that was reported but keep a common plot.name for all records coming from the same location ?

2. measurement.ID 9714, 9729, 9758, 9759 end up looking like the table below. Is that ok or do we want to "merge" "s,S" into "S" ?

ID|measurement.ID|sites.sitename|plot.name|stand.age|variable.name|date|start.date|end.date|conflicts|S.group|conflict.type
--|--|--|--|--|--|--|--|--|--|--|--
1|9714|Morgan Monroe State Forest (MMSF)|Managed|70|R_soil_C|NA|1998|1999.99726|s,S|156,157|M,T
2|9729|Morgan Monroe State Forest (MMSF)|Managed|74|R_soil_C|NI|NI|NI|s,S|156,157|M,T
3|9758|Morgan Monroe State Forest (MMSF)|Managed|70|R_soil_C|1998|NA|NA|s|156,157|M,T
4|9759|Morgan Monroe State Forest (MMSF)|Managed|71|R_soil_C|1999|NA|NA|s|156,157|M,T


3. There is a few records that you put as "S.group" in the Measurements table, as an example. But they look like 1-to-1 conflict and my code puts them as Duplicates. Should I ignore your S code for those?
example (measurement.ID 100 and 102).

Thanks!

**Comment by teixeirak** [comment]
1. This is okay. That is currently the only duplicate of that type that I am aware of. It is possible that there are some others, but your code does not need to deal with those.

**Comment by teixeirak** [comment]
2. Checked and fixed these records. The issue is removed.

**Comment by teixeirak** [comment]
3. Yes, go with your code. My entries were before we worked out exactly how all of these categories were defined.

**Comment by ValentineHerr** [comment]
For #2, since we know stand age, we could gate the date ?

**Comment by ValentineHerr** [comment]
never mind, I updated the code to deal with that.

**Comment by ValentineHerr** [comment]
Closing this issue.
I'll wrap up this and push changes in Measurements table so that you can work on the D.precedence.

**Comment by teixeirak** [comment]
Regarding your earlier comment-- In theory, yes, we can know measurement year based on stand age when we have stand age and measurement years for other records. However, I looked up this particular record and couldn't trace its origin (i.e., it didn't match original pub).

**Comment by ValentineHerr** [comment]
Ok... actually I think I should remove what I just did... because stand.age is not really "true" when it is for a record with start and end date... 
so now, 2 looks like the table below, even though we know stand age.

ID|sites.sitename|plot.name|stand.age|variable.name|date|start.date|end.date|conflicts|S.group|conflict.type
--|--|--|--|--|--|--|--|--|--|--
1|Morgan Monroe State Forest (MMSF)|Managed|70|R_soil_C|NA|1998|1999.99726|s|156,157|M,T
2|Morgan Monroe State Forest (MMSF)|Managed|74|R_soil_C|NI|NI|NI|S|156,157|M,T


Do you want me to flag those records where dates can be filled in looking at stand age ?

**Comment by teixeirak** [comment]
Okay, good, except that I ended up deleting the second record shown there (pushed new measurements file several minutes ago). That second one doesn't match up with the original pub, which is why it's such a strange duplicate.

**Comment by ValentineHerr** [comment]
measurement.ID to check:

- [ ] 5205, 5206, 5209, 5207 and 5208

**Comment by teixeirak** [comment]
I just fixed these records.
